In [42]:
import pandas as pd
import numpy as np
import os
import joblib as jb
from datetime import datetime
import src.raw_preprocessing as rp

# Multiple steps

## Conso (py pipeline)

In [2]:
conso2 = conso_preprocess("data/conso/")
conso2

,Date,Heures,Consommation
0,2021-01-01,00:00,67156.0
2,2021-01-01,00:30,66483.0
4,2021-01-01,01:00,64390.0
6,2021-01-01,01:30,64223.0
8,2021-01-01,02:00,63689.0
...,...,...,...
140249,2024-12-31,21:30,63946.0
140251,2024-12-31,22:00,63063.0
140253,2024-12-31,22:30,63816.0
140255,2024-12-31,23:00,65437.0


## School holidays (py pipeline)

In [3]:
holidays2 = school_holidays_preprocess(conso2, PATH_HOLIDAYS="data/calendar/", PATH_ARTIFACTS="artifacts/py_artifacts/")
holidays2

(Dictionnary successfully loaded)


,Date,Zone_A,Zone_B,Zone_C,Vacances de la Toussaint,Vacances de Noël,Vacances d'Hiver,Vacances de Printemps,Vacances d'Été
0,2021-01-01,True,True,True,0,1,0,0,0
2,2021-01-01,True,True,True,0,1,0,0,0
4,2021-01-01,True,True,True,0,1,0,0,0
6,2021-01-01,True,True,True,0,1,0,0,0
8,2021-01-01,True,True,True,0,1,0,0,0
...,...,...,...,...,...,...,...,...,...
140249,2024-12-31,False,False,False,0,0,0,0,0
140251,2024-12-31,False,False,False,0,0,0,0,0
140253,2024-12-31,False,False,False,0,0,0,0,0
140255,2024-12-31,False,False,False,0,0,0,0,0


## Merging of `conso` and `holidays2`

In [4]:
# We merge them with pd.concat instead of pd.merge beacause the axis and the sizes are identical.
conso2 = pd.concat([conso2, holidays2.drop("Date", axis=1)], axis=1)

In [5]:
conso2.head()

,Date,Heures,Consommation,Zone_A,Zone_B,Zone_C,Vacances de la Toussaint,Vacances de Noël,Vacances d'Hiver,Vacances de Printemps,Vacances d'Été
0,2021-01-01,00:00,67156.0,True,True,True,0,1,0,0,0
2,2021-01-01,00:30,66483.0,True,True,True,0,1,0,0,0
4,2021-01-01,01:00,64390.0,True,True,True,0,1,0,0,0
6,2021-01-01,01:30,64223.0,True,True,True,0,1,0,0,0
8,2021-01-01,02:00,63689.0,True,True,True,0,1,0,0,0


## Public Holidays (pipeline)

In [6]:
public_holidays_preprocess(conso2, "data/calendar/")
conso2

,Date,Heures,Consommation,Zone_A,Zone_B,Zone_C,Vacances de la Toussaint,Vacances de Noël,Vacances d'Hiver,Vacances de Printemps,Vacances d'Été,public_holidays
0,2021-01-01,00:00,67156.0,True,True,True,0,1,0,0,0,1
2,2021-01-01,00:30,66483.0,True,True,True,0,1,0,0,0,1
4,2021-01-01,01:00,64390.0,True,True,True,0,1,0,0,0,1
6,2021-01-01,01:30,64223.0,True,True,True,0,1,0,0,0,1
8,2021-01-01,02:00,63689.0,True,True,True,0,1,0,0,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...
140249,2024-12-31,21:30,63946.0,False,False,False,0,0,0,0,0,0
140251,2024-12-31,22:00,63063.0,False,False,False,0,0,0,0,0,0
140253,2024-12-31,22:30,63816.0,False,False,False,0,0,0,0,0,0
140255,2024-12-31,23:00,65437.0,False,False,False,0,0,0,0,0,0


# All in one

In [2]:
conso = None
if os.path.isfile("artifacts/data_artifacts/conso.pkl"):
    conso = jb.load("artifacts/data_artifacts/conso.pkl")
    print("conso dataset succesfully loaded")
else : 
    conso = rp.conso_preprocess("data/conso/")
    holidays2 = rp.school_holidays_preprocess(conso, PATH_HOLIDAYS="data/calendar/", PATH_ARTIFACTS="artifacts/py_artifacts/")
    conso = pd.concat([conso, holidays2.drop("Date", axis=1)], axis=1)
    rp.public_holidays_preprocess(conso, "data/calendar/")
    jb.dump(conso, "artifacts/data_artifacts/conso.pkl")
    print("conso dataset succesfully created")


conso dataset succesfully loaded


In [3]:
print(conso["Heures"].iloc[0])
print(conso["Date"].iloc[0])

00:00:00
2021-01-01


In [4]:
conso

,Date,Heures,Consommation,Zone_A,Zone_B,Zone_C,Vacances de la Toussaint,Vacances de Noël,Vacances d'Hiver,Vacances de Printemps,Vacances d'Été,public_holidays
0,2021-01-01,00:00:00,67156.0,True,True,True,0,1,0,0,0,0
2,2021-01-01,00:30:00,66483.0,True,True,True,0,1,0,0,0,0
4,2021-01-01,01:00:00,64390.0,True,True,True,0,1,0,0,0,0
6,2021-01-01,01:30:00,64223.0,True,True,True,0,1,0,0,0,0
8,2021-01-01,02:00:00,63689.0,True,True,True,0,1,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...
140249,2024-12-31,21:30:00,63946.0,False,False,False,0,0,0,0,0,0
140251,2024-12-31,22:00:00,63063.0,False,False,False,0,0,0,0,0,0
140253,2024-12-31,22:30:00,63816.0,False,False,False,0,0,0,0,0,0
140255,2024-12-31,23:00:00,65437.0,False,False,False,0,0,0,0,0,0


In [27]:
meteo75 = pd.read_csv("data/weather/H_75_previous-2020-2024.csv", sep =';')

In [28]:
meteo75

,NUM_POSTE,NOM_USUEL,LAT,LON,ALTI,AAAAMMJJHH,RR1,QRR1,DRR1,QDRR1,...,INS2,QINS2,TLAGON,QTLAGON,TVEGETAUX,QTVEGETAUX,ECOULEMENT,QECOULEMENT,STATUS_FXI3S,STATUS_DXI3S
0,75106001,LUXEMBOURG,48.844833,2.338500,50,2020010100,0.0,1.0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,75106001,LUXEMBOURG,48.844833,2.338500,50,2020010101,0.0,1.0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,75106001,LUXEMBOURG,48.844833,2.338500,50,2020010102,0.0,1.0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,75106001,LUXEMBOURG,48.844833,2.338500,50,2020010103,0.0,1.0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,75106001,LUXEMBOURG,48.844833,2.338500,50,2020010104,0.0,1.0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
261855,75116008,LONGCHAMP,48.854833,2.233667,27,2024123119,0.0,1.0,NaN,NaN,...,0.0,9.0,NaN,NaN,NaN,NaN,NaN,NaN,0.0,1.0
261856,75116008,LONGCHAMP,48.854833,2.233667,27,2024123120,0.0,1.0,NaN,NaN,...,0.0,9.0,NaN,NaN,NaN,NaN,NaN,NaN,0.0,1.0
261857,75116008,LONGCHAMP,48.854833,2.233667,27,2024123121,0.0,1.0,NaN,NaN,...,0.0,9.0,NaN,NaN,NaN,NaN,NaN,NaN,0.0,1.0
261858,75116008,LONGCHAMP,48.854833,2.233667,27,2024123122,0.0,1.0,NaN,NaN,...,0.0,9.0,NaN,NaN,NaN,NaN,NaN,NaN,0.0,1.0


In [29]:
cols_to_keep = [
    'NUM_POSTE',        # station ID : identifies the city
    'LAT', 'LON',       # coordinates: used to query Open-Meteo
    'AAAAMMJJHH',       # timestamp
    # Usefull features that we can found with the API of open-meteo
    'T',                # temperature_2m        
    'U',                # relative_humidity_2m  
    'FF',               # wind_speed_10m        
    'PMER',             # pressure_msl          
    'RR1',              # precipitation         
]

meteo75 = meteo75[cols_to_keep]

In [30]:
meteo75

,NUM_POSTE,LAT,LON,AAAAMMJJHH,T,U,FF,PMER,RR1
0,75106001,48.844833,2.338500,2020010100,1.6,NaN,NaN,NaN,0.0
1,75106001,48.844833,2.338500,2020010101,0.9,NaN,NaN,NaN,0.0
2,75106001,48.844833,2.338500,2020010102,0.1,NaN,NaN,NaN,0.0
3,75106001,48.844833,2.338500,2020010103,0.2,NaN,NaN,NaN,0.0
4,75106001,48.844833,2.338500,2020010104,0.6,NaN,NaN,NaN,0.0
...,...,...,...,...,...,...,...,...,...
261855,75116008,48.854833,2.233667,2024123119,4.9,94.0,5.0,NaN,0.0
261856,75116008,48.854833,2.233667,2024123120,5.3,93.0,5.1,NaN,0.0
261857,75116008,48.854833,2.233667,2024123121,5.6,93.0,4.4,NaN,0.0
261858,75116008,48.854833,2.233667,2024123122,5.6,95.0,5.0,NaN,0.0


#### Let's separate the date and hours by creating 2 new columns and make them compatible with the main dataset `conso`

In [31]:
meteo75["Date"] = meteo75["AAAAMMJJHH"].apply(lambda x : str(str(x)[:8]))
meteo75["Date"] = meteo75["Date"].apply(lambda x : datetime.strptime(x, '%Y%m%d').date())

meteo75["Heures"] = meteo75["AAAAMMJJHH"].apply(lambda x : str(str(x)[8:]) + ':00')
meteo75["Heures"] = meteo75["Heures"].apply(lambda x : datetime.strptime(x, '%H:%M').time())

meteo75 = meteo75.drop("AAAAMMJJHH", axis=1)

In [32]:
meteo75

,NUM_POSTE,LAT,LON,T,U,FF,PMER,RR1,Date,Heures
0,75106001,48.844833,2.338500,1.6,NaN,NaN,NaN,0.0,2020-01-01,00:00:00
1,75106001,48.844833,2.338500,0.9,NaN,NaN,NaN,0.0,2020-01-01,01:00:00
2,75106001,48.844833,2.338500,0.1,NaN,NaN,NaN,0.0,2020-01-01,02:00:00
3,75106001,48.844833,2.338500,0.2,NaN,NaN,NaN,0.0,2020-01-01,03:00:00
4,75106001,48.844833,2.338500,0.6,NaN,NaN,NaN,0.0,2020-01-01,04:00:00
...,...,...,...,...,...,...,...,...,...,...
261855,75116008,48.854833,2.233667,4.9,94.0,5.0,NaN,0.0,2024-12-31,19:00:00
261856,75116008,48.854833,2.233667,5.3,93.0,5.1,NaN,0.0,2024-12-31,20:00:00
261857,75116008,48.854833,2.233667,5.6,93.0,4.4,NaN,0.0,2024-12-31,21:00:00
261858,75116008,48.854833,2.233667,5.6,95.0,5.0,NaN,0.0,2024-12-31,22:00:00


Let's check that if all the timeframes are there

In [33]:
meteo75[meteo75["NUM_POSTE"]==75106001]["Heures"].unique()

array([datetime.time(0, 0), datetime.time(1, 0), datetime.time(2, 0),
       datetime.time(3, 0), datetime.time(4, 0), datetime.time(5, 0),
       datetime.time(6, 0), datetime.time(7, 0), datetime.time(8, 0),
       datetime.time(9, 0), datetime.time(10, 0), datetime.time(11, 0),
       datetime.time(12, 0), datetime.time(13, 0), datetime.time(14, 0),
       datetime.time(15, 0), datetime.time(16, 0), datetime.time(17, 0),
       datetime.time(18, 0), datetime.time(19, 0), datetime.time(20, 0),
       datetime.time(21, 0), datetime.time(22, 0), datetime.time(23, 0)],
      dtype=object)

#### We'll keep the rows where the date is between `min(conso["Date"])` and `max(conso["Date"])`

In [34]:
min_date = conso["Date"].iloc[0]
max_date = conso["Date"].iloc[-1]

meteo75 = meteo75[
    (meteo75["Date"]>= min_date) & 
    (meteo75["Date"]<= max_date)
    ]

In [35]:
meteo75

,NUM_POSTE,LAT,LON,T,U,FF,PMER,RR1,Date,Heures
8784,75106001,48.844833,2.338500,0.9,NaN,NaN,NaN,0.0,2021-01-01,00:00:00
8785,75106001,48.844833,2.338500,0.5,NaN,NaN,NaN,0.0,2021-01-01,01:00:00
8786,75106001,48.844833,2.338500,0.8,NaN,NaN,NaN,0.0,2021-01-01,02:00:00
8787,75106001,48.844833,2.338500,0.4,NaN,NaN,NaN,0.0,2021-01-01,03:00:00
8788,75106001,48.844833,2.338500,0.7,NaN,NaN,NaN,0.0,2021-01-01,04:00:00
...,...,...,...,...,...,...,...,...,...,...
261855,75116008,48.854833,2.233667,4.9,94.0,5.0,NaN,0.0,2024-12-31,19:00:00
261856,75116008,48.854833,2.233667,5.3,93.0,5.1,NaN,0.0,2024-12-31,20:00:00
261857,75116008,48.854833,2.233667,5.6,93.0,4.4,NaN,0.0,2024-12-31,21:00:00
261858,75116008,48.854833,2.233667,5.6,95.0,5.0,NaN,0.0,2024-12-31,22:00:00


### We can observe that there are multiple stations / data providers.  
There are differents weather stations so there are several duplicate rows for a same date

In [14]:
print("There are", len(meteo75["NUM_POSTE"].unique()), "stations")

There are 6 stations


### We guess that some stations didn't record all the weather data, but that others did.  
Let's calculate the percentage of missing values for each columns and also grouped each station

In [94]:
print("Percentage of NaN per column")
print((meteo75.isnull().mean() * 100).round(1))

print("\nPercentage of NaN per station")
print(meteo75.groupby('NUM_POSTE')[['T', 'U', 'FF', 'PMER', 'RR1']].apply(
    lambda x: x.isnull().mean() * 100))

Percentage of NaN per column
NUM_POSTE     0.0
LAT           0.0
LON           0.0
T             0.2
U            66.5
FF           50.4
PMER         83.2
RR1          33.3
Date          0.0
Heures        0.0
dtype: float64

Percentage of NaN per station
                  T      U          FF   PMER         RR1
NUM_POSTE                                                
75106001   0.008556  100.0  100.000000  100.0    0.444901
75107005   1.182068  100.0    0.390082  100.0  100.000000
75110001   0.000000  100.0  100.000000  100.0    0.410678
75114001   0.000000    0.0    0.011408    0.0    0.002852
75114007   0.037076  100.0  100.000000  100.0  100.000000
75116008   0.077002    0.0    0.322268  100.0    1.026694


We are going to keep the station that has the fewest missing values

In [57]:
grid_scores = meteo75.groupby('NUM_POSTE')[['T', 'U', 'FF', 'PMER', 'RR1']].apply(
    lambda x: x.isnull().mean() * 100)

best_station = grid_scores.sum(axis=1).idxmin()
print("Station n° ", best_station, " Sum percentages: ", min(grid_scores.sum(axis=1)))
print("Best station : ", best_station)

Station n°  75114001  Sum percentages:  0.014259639516313028
Best station :  75114001


For example, we can observe here that the station `75114001` has recorded more data than the others.  

In [62]:
grid_scores_array = np.array(grid_scores) 
x, y = grid_scores_array.shape

In [63]:
x, y

(6, 5)

In [75]:
weights = np.array([3, 2, 2, 1, 1]).reshape((y, 1))
best_station_idx = grid_scores_array.dot(weights).argmin()
int(grid_scores.iloc[best_station_idx].name)

75114001

In [17]:
meteo75[meteo75["NUM_POSTE"] == best_station].isnull().mean() * 100

NUM_POSTE    0.000000
LAT          0.000000
LON          0.000000
T            0.000000
U            0.000000
FF           0.011408
PMER         0.000000
RR1          0.002852
Date         0.000000
Heures       0.000000
dtype: float64

In [18]:
meteo75 = meteo75[meteo75["NUM_POSTE"] == best_station]

In [41]:
meteo75

,NUM_POSTE,LAT,LON,T,U,FF,PMER,RR1,Date,Heures
139101,75114001,48.821667,2.337833,1.6,84.0,2.0,1008.0,0.0,2021-01-01,00:00:00
139102,75114001,48.821667,2.337833,1.4,84.0,1.8,1008.2,0.0,2021-01-01,01:00:00
139103,75114001,48.821667,2.337833,0.9,86.0,1.6,1008.6,0.0,2021-01-01,02:00:00
139104,75114001,48.821667,2.337833,0.8,87.0,1.8,1008.9,0.0,2021-01-01,03:00:00
139105,75114001,48.821667,2.337833,0.6,87.0,1.1,1009.0,0.0,2021-01-01,04:00:00
...,...,...,...,...,...,...,...,...,...,...
174160,75114001,48.821667,2.337833,4.5,94.0,3.5,1028.7,0.0,2024-12-31,19:00:00
174161,75114001,48.821667,2.337833,4.7,93.0,3.6,1028.7,0.0,2024-12-31,20:00:00
174162,75114001,48.821667,2.337833,5.1,93.0,4.8,1028.2,0.0,2024-12-31,21:00:00
174163,75114001,48.821667,2.337833,5.2,93.0,6.9,1028.3,0.0,2024-12-31,22:00:00


In [36]:
rp.monitoring_nan(meteo75)

There are 6 stations
Percentage of NaN per column
NUM_POSTE     0.0
LAT           0.0
LON           0.0
T             0.2
U            66.5
FF           50.4
PMER         83.2
RR1          33.3
Date          0.0
Heures        0.0
dtype: float64

Percentage of NaN per station
                  T      U          FF   PMER         RR1
NUM_POSTE                                                
75106001   0.008556  100.0  100.000000  100.0    0.444901
75107005   1.182068  100.0    0.390082  100.0  100.000000
75110001   0.000000  100.0  100.000000  100.0    0.410678
75114001   0.000000    0.0    0.011408    0.0    0.002852
75114007   0.037076  100.0  100.000000  100.0  100.000000
75116008   0.077002    0.0    0.322268  100.0    1.026694


In [43]:
meteo75 = rp.select_best_station(meteo75, conso)

The best station has a weighted average of :  [[0.02566735]]
Best station :  75114001


In [45]:
rp.monitoring_nan(meteo75)

There are 1 stations
Percentage of NaN per column
NUM_POSTE    0.0
LAT          0.0
LON          0.0
T            0.0
U            0.0
FF           0.0
PMER         0.0
RR1          0.0
Date         0.0
Heures       0.0
dtype: float64

Percentage of NaN per station
             T    U        FF  PMER       RR1
NUM_POSTE                                    
75114001   0.0  0.0  0.011408   0.0  0.002852


#### For the missing values we'll use the interpolation method  
The interpolation method estimates missing values by fitting a function bewteen two known neighboring points and evaluating it at the missing timestamp (assuming the variable evolves smoothly and continuously bewteen two observations).  

We'll use the function below to estimate the missing value, like temperature, from a timestamp between the known values from its 2 neighboring timestamps:  

> $T(t) = T_{1} + (\frac{T_{2}-T_{1}}{t_{2}-t_{1}}) *  (t-t_{1})$

We calculate the slope $\alpha$ of the function that goes through two known points, $T_{1}$ and $T_{2}$, from two timestamps $t_{1}$ and $t_{2}$ as close together as possible :  
> $ \alpha = \frac{T_{2} - T_{1}}{t_{2}-t_{1}}$

We assume a linear relationship between these 2 points, so the affine function will look like this :  
> $f(t) = \alpha * t + \beta$,  
> with $ \beta = f(t) -\alpha*t$

We know the value $T_{1} = f(t_{1})$ for a surrounding timestamp $t_{1}$.  
So : 
> $\beta = T_{1} -\alpha*t_{1}$  

We could've chosen the other surrounding timestamp $t_{2}$ and its value $T_{2}$ instead of $t_{1}$ and $T_{1}$.

Then we substitute $\beta$ with ($T_{1} -\alpha*t_{1}$) : 
> $f(t) = \alpha * t + (T_{1} -\alpha*t_{1})$  
> $f(t) = T_{1} + \alpha * (t - t_{1})$  
> $f(t) = T_{1} + (\frac{T_{2}-T_{1}}{t_{2}-t_{1}}) *  (t-t_{1})$

In [48]:
def interpolation(x, n1, n2):
    x1, X1 = n1
    x2, X2 = n2
    return X1 + (x-x1)*(X2-X1)/(x2-x1)

In [86]:
pd.isna(meteo75["FF"].iloc[0])

False

In [118]:
def time_to_float(t):
    """Convert datetime.time to float hours 
    (example :  14:30 -> 14.5)
    """
    return t.hour + t.minute / 60


def interpolation_col(df):
    df = df.reset_index(drop=True)
    columns = list(set(df.columns) - set(["NUM_POSTE", "LAT", "LON", "Date", "Heures"]))
    
    for col in columns:
        # If the column has Nan Values, we apply the interpolation method on it
        if (meteo75[col].isnull().mean() * 100) != 0.0 :
            for i in range(len(df)):
                if(pd.isna(df[col].iloc[i])):
                    # If the missing value is in the first row, we choose the value of the next row
                    if(i == 0):
                        df[i, col] = df[col].iloc[i+1]
                    # If the missing value is in the last row, we choose the value of the last row
                    elif i == len(df)-1 : 
                        df[i, col] = df[col].iloc[i-1]
                    else : 
                        t = time_to_float(df["Heures"].iloc[i])
                        t1, T1 = time_to_float(df["Heures"].iloc[i-1]), df[col].iloc[i-1]    
                        # We select the index of the next valid row (Not NaN value)
                        # first_valid_index returns an index, not a position
                        idx_next = df[col].iloc[i+1:].first_valid_index()
                        if idx_next is None:
                            df.loc[i, col] = T1
                            continue

                        t2, T2 = time_to_float(df["Heures"].loc[idx_next]), df[col].loc[idx_next]
                        
                        df.loc[i, col] = interpolation(t, (t1, T1), (t2, T2))
    return df
                    

In [119]:
meteo75 = interpolation_col(meteo75)

In [123]:
meteo75.info()

<class 'pandas.DataFrame'>
RangeIndex: 35064 entries, 0 to 35063
Data columns (total 10 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   NUM_POSTE  35064 non-null  int64  
 1   LAT        35064 non-null  float64
 2   LON        35064 non-null  float64
 3   T          35064 non-null  float64
 4   U          35064 non-null  float64
 5   FF         35064 non-null  float64
 6   PMER       35064 non-null  float64
 7   RR1        35064 non-null  float64
 8   Date       35064 non-null  object 
 9   Heures     35064 non-null  object 
dtypes: float64(7), int64(1), object(2)
memory usage: 2.7+ MB


#### Now the Nan values are filled

In [122]:
print("Percentage of NaN per column")
print((meteo75.isnull().mean() * 100))

Percentage of NaN per column
NUM_POSTE    0.0
LAT          0.0
LON          0.0
T            0.0
U            0.0
FF           0.0
PMER         0.0
RR1          0.0
Date         0.0
Heures       0.0
dtype: float64


#### We'll use the pandas built-in method to interpolate the missing values